lib/Wikia/src/Util/SamplerProxy.php

```diff
@@ -0,0 +1,154 @@
+ *  6, Gradually increase sampling percentage until the new service/functionality handles 100% of
+ *     live calls without issues.
+ *  7. Deploy a code change that removes the SamplerProxy and directly calls the new
+ *     service/functionality and deletes the old code.
+ */
+namespace Wikia\Util;
+
+
+class SamplerProxy {
+
+	/**
+	 * @var string
+	 */
+	protected $enableShadowing;
+
+	/**
+	 * @var string
+	 */
+	protected $methodSamplingRate;
+
+	/**
+	 * @var callable
+	 * Expected to be of the form [ <instance>, <method name> ]
+	 */
+	protected $originalCallable;
+
+	/**
+	 * @var callable
+	 * Expected to be of the form [ <instance>, <method name> ]
+	 * Should have the same argument list as $originalCallable
+	 */
+	protected $alternateCallable;
+
+	/**
+	 * @var callable
+	 * Takes 2 arguments, ( $originalResults, $alternateResults ) and
+	 * returns the results that should be passed back to the caller
+	 */
+	protected $resultsCallable;
+
+	public static function createBuilder() {
+		return new SamplerProxyBuilder();
+	}
+
+	public function __construct( SamplerProxyBuilder $samplerProxyBuilder ) {
+		$this->enableShadowing = $samplerProxyBuilder->getEnableShadowing();
+		$this->methodSamplingRate = $samplerProxyBuilder->getMethodSamplingRate();
+		$this->originalCallable = $samplerProxyBuilder->getOriginalCallable();
+		$this->alternateCallable = $samplerProxyBuilder->getAlternateCallable();
+		$this->resultsCallable = $samplerProxyBuilder->getResultsCallable();
+	}
+
+	public function __call( $method, $args ) {
+
+		// Forward all methods but $originalMethod to $originalPopo
+		if ( $method != $this->originalCallable[1] ) {
+			return call_user_func_array( [ $this->originalCallable[0], $method ], $args );
+		}
+
+		if ( !empty( $this->methodSamplingRate ) && rand( 0, 100 ) <= $this->methodSamplingRate ) {
+
+			// We're going to route the request to the alternate instance.
+			// Now determine if we're going to shadow or redirect
+			$shouldShadow = $this->enableShadowing;
+
+			// Call the alternate method and cache the result
+			try {
+				$alternateResult = call_user_func_array( $this->alternateCallable, $args );
+			}
+			catch ( \Exception $e ) {
+				$log = \Wikia\Logger\WikiaLogger::instance();
+				$log->error( __METHOD__ . " - Exception thrown from alternate method: " .
+				             $e->getMessage(), [
+					'method' => __METHOD__,
+					'exception' => $e,
+					'shadowing' => $this->enableShadowing,
+					'sampleRate' => $this->methodSamplingRate,
+				] );
+
+				// An exception was thrown by the alternate method.  Call the original method in
+				// an attempt to obtain valid results for the caller.  This seems like a reasonable
+				// fallback during a temporary migration scenario as should always be the case when
+				// you are using this proxy object.
+				$shouldShadow = true;
+			}
+
+			if ( !$shouldShadow && isset( $alternateResult ) ) {
+				$resultToReturn = $alternateResult;
+			} else {
```

Comment:
It is a possible state and it is checked below on line 115.  At this point we haven't returned the alternate result so we're committed to at least calling the original, hence the else.